// nodemailer-browser/Gruntfile.js
'use strict';

var fs = require('fs-extra');

module.exports = function(grunt) {

  grunt.loadNpmTasks('grunt-contrib-jshint');
  grunt.loadNpmTasks('grunt-mocha-cli');
  grunt.loadNpmTasks('grunt-release');

  grunt.initConfig({
    jshint: {
      files: ['Gruntfile.js', 'lib/index.js', 'test/index.js'],
      options: {
        jshintrc: '.jshintrc'
      }
    },
    mochacli: {
      options: {
        require: ['chai'],
        reporter: 'spec',
        timeout: 10000
      },
      test: {
        options: {
          files: ['test/connect-dynamic-middleware.js']
        }
      }
    },
    release: {
      options: {
        tagName: 'v<%= version %>'
      }
    }
  });

  grunt.registerTask('clean-environment', 'remove temp folder for tests', function() {
    fs.removeSync('./tmp');
  });

  grunt.registerTask('test', ['jshint', 'gjslint', 'mochacli:test', 'clean-environment']);
  grunt.registerTask('linters', ['jshint', 'gjslint']);
  grunt.registerTask('default', ['test']);

  grunt.registerTask('gjslint', 'run the closure linter', function() {
    var done = this.async();
    var args = [
      './scripts/gjslint.py',
      '--disable',
      '0110',
      '--nojsdoc',
      '-r',
      '.',
      '-e',
      'node_modules'
    ];

    var child = require('child_process').spawn('python', args);
    child.stdout.on('data', function(chunk) { grunt.log.write(chunk); });
    child.stderr.on('data', function(chunk) { grunt.log.error(chunk); });
    child.on('close', function(code) { done(code ? false : true); });
  });

};
